src/sse/sse_butterflies.rs

```diff
@@ -0,0 +1,4246 @@
+        value3: __m128d,
+        value4: __m128d,
+    ) -> [__m128d; 5] {
+        // This is a SSE translation of the scalar 5-point butterfly
+        let x14p = _mm_add_pd(value1, value4);
+        let x14n = _mm_sub_pd(value1, value4);
+        let x23p = _mm_add_pd(value2, value3);
+        let x23n = _mm_sub_pd(value2, value3);
+
+        let temp_a1_1 = _mm_mul_pd(self.twiddle1re, x14p);
+        let temp_a1_2 = _mm_mul_pd(self.twiddle2re, x23p);
+        let temp_a2_1 = _mm_mul_pd(self.twiddle2re, x14p);
+        let temp_a2_2 = _mm_mul_pd(self.twiddle1re, x23p);
+
+        let temp_b1_1 = _mm_mul_pd(self.twiddle1im, x14n);
+        let temp_b1_2 = _mm_mul_pd(self.twiddle2im, x23n);
+        let temp_b2_1 = _mm_mul_pd(self.twiddle2im, x14n);
+        let temp_b2_2 = _mm_mul_pd(self.twiddle1im, x23n);
+
+        let temp_a1 = _mm_add_pd(value0, _mm_add_pd(temp_a1_1, temp_a1_2));
+        let temp_a2 = _mm_add_pd(value0, _mm_add_pd(temp_a2_1, temp_a2_2));
+
+        let temp_b1 = _mm_add_pd(temp_b1_1, temp_b1_2);
+        let temp_b2 = _mm_sub_pd(temp_b2_1, temp_b2_2);
+
+        let temp_b1_rot = self.rotate.rotate(temp_b1);
+        let temp_b2_rot = self.rotate.rotate(temp_b2);
+
+        let x0 = _mm_add_pd(value0, _mm_add_pd(x14p, x23p));
+        let x1 = _mm_add_pd(temp_a1, temp_b1_rot);
+        let x2 = _mm_add_pd(temp_a2, temp_b2_rot);
+        let x3 = _mm_sub_pd(temp_a2, temp_b2_rot);
+        let x4 = _mm_sub_pd(temp_a1, temp_b1_rot);
+        [x0, x1, x2, x3, x4]
+    }
+}
+
+//    __             _________  _     _ _
+//   / /_           |___ /___ \| |__ (_) |_
+//  | '_ \   _____    |_ \ __) | '_ \| | __|
+//  | (_) | |_____|  ___) / __/| |_) | | |_
+//   \___/          |____/_____|_.__/|_|\__|
+//
+
+pub struct SseF32Butterfly6<T> {
+    direction: FftDirection,
+    _phantom: std::marker::PhantomData<T>,
+    bf3: SseF32Butterfly3<T>,
+}
+
+boilerplate_fft_sse_f32_butterfly!(SseF32Butterfly6, 6, |this: &SseF32Butterfly6<_>| this
+    .direction);
+boilerplate_fft_sse_common_butterfly!(SseF32Butterfly6, 6, |this: &SseF32Butterfly6<_>| this
+    .direction);
+impl<T: FftNum> SseF32Butterfly6<T> {
+    #[inline(always)]
+    pub fn new(direction: FftDirection) -> Self {
+        assert_f32::<T>();
+        let bf3 = SseF32Butterfly3::new(direction);
+
+        Self {
+            direction,
+            _phantom: std::marker::PhantomData,
+            bf3,
+        }
+    }
+
+    #[inline(always)]
+    pub(crate) unsafe fn perform_fft_contiguous(
+        &self,
+        input: RawSlice<Complex<T>>,
+        output: RawSliceMut<Complex<T>>,
+    ) {
+        let value01 = _mm_loadu_ps(input.as_ptr() as *const f32);
+        let value23 = _mm_loadu_ps(input.as_ptr().add(2) as *const f32);
+        let value45 = _mm_loadu_ps(input.as_ptr().add(4) as *const f32);
+
+        let temp = self.perform_fft_direct(value01, value23, value45);
+
+        let array = std::mem::transmute::<[__m128; 3], [Complex<f32>; 6]>(temp);
```

Comment:
Unfortunately, I'm pretty sure transmuting vector types like __m128 as arrays of scalar types is UB.

Reply:
I can't find anything in the docs about this. How is this done in the avx code?

Reply:
Most AVX butterflies follow the six-step FFT platter, but can skip the first and last transpose.

The six step FFT algorithm is:
1: Transpose the data
2: Compute contiguous FFTs down the rows of our transposed array. These rows were originally the input columns, but became rows in the transpose.
3: Apply twiddle factors
4: Transpose
5: Compute contiguous FFTs down the rows of the array we transposed. These rows will eventually become the columns of the output.
6: Transpose into the output.

If you skip step #1, then in #2, you do FFTs down the columns of the raw input array, without having to do any preprocessing on it - just load data and start FFTing. This is why they're called column FFTs -- because they're doing FFTs down the columns of the input data. And likewise, if you skip step #6, then you do more column FFTs in #5, and the output of the column FFTs can be written directly to memory with no postprocessing needed.

The clearest, most concise example is probably the AVX f32 size-16 butterfly:

```rust
        let rows = [
            input.load_complex(0),
            input.load_complex(4),
            input.load_complex(8),
            input.load_complex(12),
        ];

        // We're going to treat our input as a 4x4 2d array. First, do 4 butterfly 4's down the columns of that array.
        let mut mid = AvxVector::column_butterfly4(rows, self.twiddles_butterfly4);

        // apply twiddle factors
        for r in 1..4 {
            mid[r] = AvxVector::mul_complex(mid[r], self.twiddles[r - 1]);
        }

        // Transpose our 4x4 array
        let transposed = avx32_utils::transpose_4x4_f32(mid);

        // Do 4 butterfly 4's down the columns of our transposed array
        let output_rows = AvxVector::column_butterfly4(transposed, self.twiddles_butterfly4);

        output.store_complex(output_rows[0], 0);
        output.store_complex(output_rows[1], 4);
        output.store_complex(output_rows[2], 8);
        output.store_complex(output_rows[3], 12);
```
More generally, in AVX butterflies, the process goes like this:
1: Compute column butterflies down the columns of the input
2: Apply twiddle factors
3: Transpose
4: Compute column butterflies again on the transposed data

If you switched to this model, it'd be more convenient across the board to load and store, and you could store the AVX vectors directly into the output instead of the middle step of transmuting.

Reply:
Note that if you switch to this model for butterflies that use good-thomas algorithm, you'll hit a snag that you can't do the good-thomas reordering, because that's preprocessing, and this model is designed to not need preprocessing.

It's worth looking into
1. For direct (not dual) butterflies, is it better to just apply twiddle factors? This neded up being the case for a few of the AVX butterflies. It'd be a single call to `complex_mul` in this case, because you can process the first column separately, which doesn't need butterflies, and then pack the second and third columns into a single vector.
2. Alternatively, I discovered during AVX work that it's al ways possible to find a good-thomas mapping that keeps each element in its original column, and it works out that you can do it with one shuffle per row, on both input and output. That would defeat the benefits of the "truncated six step" model though.
3. If we upgrade to SSE 4.1 instead of sticking to SSE3, we can use `_mm_blend_ps` https://software.intel.com/sites/landingpage/IntrinsicsGuide/#techs=SSE4_1&expand=447 Blend is like a shuffle, but it's much faster, and it keeps everything in the same column as the inputs - which like I mentioned in #2, is a constraint that this algorithm is suited for. So we could use blends to accomplish the input reordering, do the column FFTs, transpose, do more column FFTs, then use blends to accomplish the output reordering.

Reply:
The transmute isn't needed any more, after I switched to the SseArray trait.